lib/ReactGridLayout.jsx

```diff
@@ -290,20 +301,20 @@ export default class ReactGridLayout extends React.Component<Props, State> {
     this.onLayoutMaybeChanged(this.state.layout, this.props.layout);
   }
 
-  componentWillReceiveProps(nextProps: Props) {
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
```

Comment:
There is a really big performance impact with those changes. I'm getting lag, since when we use `getDerivedStateFromProps` and `componentDidUpdate` it does the equality checks every single time. They're not equivalent to `componentWillReceiveProps` which is only triggered when *props* change (not state). Not sure how we can optimize this.

Also if we go this route, we can do the changes without `getDerivedStateFromProps` and just go with `componentDidUpdate` (with adding a check of what changed and trigger state changes only when needed).

Reply:
I did some changes locally on my branch very similar to yours, but it impacts performance (maybe we can do some measurements and see how bad the impact actually is). Nevertheless, I think for such big changes it will be a good idea to involve @STRML and see what he thinks.

Reply:
Yeah this is problematic, this whole block is going to run many times while dragging an item. We'll need to profile & brainstorm where we can optimize.

Reply:
@STRML @n1ghtmare we can try this

> Also if we go this route, we can do the changes without getDerivedStateFromProps and just go with componentDidUpdate

with props mirroring. It can help us to avoid extra comparisons

What do you think?

Reply:
We may want to require the library user to send a new `layout` object (rather than mutating the existing one) or supply something optional like a `layoutDidUpdate` function of arity `(newLayout: Layout, oldLayout: Layout) => boolean` where the default implementation is `(newLayout, oldLayout) => newLayout !== oldLayout`.

This will help us avoid an expensive deep comparison.

Perhaps some of you know a way this is handled in other libraries? I don't want to require the author to mutate `key`, which will unmount all children.

Reply:
I think that layout recreating on the user side is not an obvious way to fix it. I mean not obvious for a user

Maybe replacing the code into componentDidMount will be enough to fix performance issues. I'll try to compare cases with componentWillReceiveProps, getDerivedStateFromProps + componentDidUpdate and componentDidUpdate

Reply:
@STRML Just for the sake of brainstorming (and I'm sure I'm missing something), but are we doing this comparison only for when the `compactType` changes or when an item (child) is added/removed?